### PR TITLE
Implement std::error::Error for Error Types

### LINKED ISSUE #110: Implement `std::error::Error` for error types
There are error types present that do not implement [`std::error::Error`](https://doc.rust-lang.org/std/error/trait.Error.html) e.g. https://github.com/rust-vmm/vm-superio/blob/main/vm-superio/src/serial.rs#L281:L290.

This makes error propagation more difficult in dependent crates e.g. https://github.com/firecracker-microvm/firecracker/pull/4585#pullrequestreview-2031394720.

These types should implement `std::error::Error`.

### PR DESCRIPTION
**Pull request by Ecazares15**
Updated Error<E> in serial.rs to implement std::error::Error.

### Summary of the PR

Closes #110 

### Requirements

Before submitting your PR, please make sure you addressed the following
requirements:

- [ ] All commits in this PR have Signed-Off-By trailers (with
  `git commit -s`), and the commit message has max 60 characters for the
  summary and max 75 characters for each description line.
- [ ] All added/changed functionality has a corresponding unit/integration
  test.
- [ ] All added/changed public-facing functionality has entries in the "Upcoming 
  Release" section of CHANGELOG.md (if no such section exists, please create one).
- [ ] Any newly added `unsafe` code is properly documented.

### CODE REVIEW

File: vm-superio/src/serial.rs [hunk 1]
```diff
@@ -289,6 +291,18 @@ pub enum Error<E> {
     FullFifo,
 }
 
+impl fmt::Display for Error<io::Error> {
```
Comment (roypat) [inline vm-superio/src/serial.rs:294]: Mh, not sure if `E` is ever not `io::Error`, but could we still do 
```suggestion
impl<E: fmt::Display> fmt::Display for Error<E> {
```
here?

Comment (Ecazares15) [reply vm-superio/src/serial.rs]: got it!

File: vm-superio/src/serial.rs [hunk 2]
```diff
@@ -11,6 +11,8 @@
 //! This is done by emulating an UART serial port.
 
 use std::collections::VecDeque;
+use std::error::Error as StdError;
```
Comment (JonathanWoollett-Light) [inline vm-superio/src/serial.rs:14]: I would prefer using the full path e.g. `impl<E: std::error::Error> std::error::Error for Error<E> {}`.

File: vm-superio/src/serial.rs [hunk 3]
```diff
@@ -289,6 +291,18 @@ pub enum Error<E> {
...
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            Error::Trigger(e) => write!(f, "Failed to trigger interrupt: {}", e),
+            Error::IOError(e) => write!(f, "Couldn't write/flush to the given destination: {}", e),
+            Error::FullFifo => write!(f, "No space left in FIFO"),
+        }
+    }
+}
+
+impl StdError for Error<io::Error> {}
```
Comment (JonathanWoollett-Light) [inline vm-superio/src/serial.rs:304]: This should also be generic over `E`.
```suggestion
impl<E: StdError> StdError for Error<E> {}
```

File: vm-superio/src/serial.rs (hunk 2 above)
Comment (Ecazares15) [reply vm-superio/src/serial.rs:14]: Using the full path causes a redefinition error of `Error` the suggested change was to use `as` to change the binding name of the import.

File: vm-superio/src/serial.rs (hunk 3 above)
Comment (Ecazares15) [reply vm-superio/src/serial.rs]: got it!

**Review by roypat — APPROVED** [review]

**Review by ShadowCurse — APPROVED** [review]

**Pull request merged**